db-service/lib/common/DatabaseService.js

```diff
@@ -1,143 +1,141 @@
     const ctx = this.context
-    if (!ctx) return this.tx().begin()
-    const tenant = this.isMultitenant && ctx.tenant
-    const pool = (this.pools[tenant] ??= new Pool(this.pools._factory, tenant))
-    const connections = pool._trackedConnections
-    let dbc
-    try {
-      dbc = this.dbc = await pool.acquire()
-    } catch (err) {
-      // TODO: add acquire timeout error check
-      err.stack += `\nActive connections:${connections.length}\n${connections.map(c => c._beginStack.stack).join('\n')}`
-      throw err
-    }
-    this._beginStack = new Error('begin called from:')
-    connections.push(this)
-    this._release = async dbc => {
-      await pool.release(dbc)
-      connections.splice(connections.indexOf(this), 1)
-    }
+    if (!ctx) return this.tx().begin() // REVISIT: Is this correct? When does this happen?
+
+    // REVISIT: tenant should be undefined if !this.isMultitenant
+    let isMultitenant = 'multiTenant' in this.options ? this.options.multiTenant : cds.env.requires.multitenancy
+    let tenant = isMultitenant && ctx.tenant
+
+    // Setting this.pool as used in this.acquire() and this.release()
+    this.pool = this.pools[tenant] ??= new ConnectionPool(this.pools._factory, tenant)
+
+    // Acquire a pooled connection
+    this.dbc = await this.acquire()
+
+    // Begin a session...
     try {
-      // Setting session context variables
-      await this.set({
-        get '$user.id'() {
-          return _set(this, '$user.id', ctx.user?.id || 'anonymous')
-        },
-        get '$user.locale'() {
-          return _set(this, '$user.locale', ctx.locale || cds.env.i18n.default_language)
-        },
-        get '$valid.from'() {
-          return _set(this, '$valid.from', ctx._?.['VALID-FROM'] ?? ctx._?.['VALID-AT'] ?? '1970-01-01T00:00:00.000Z')
-        },
-        get '$valid.to'() {
-          return _set(
-            this,
-            '$valid.to',
-            ctx._?.['VALID-TO'] ?? _validTo4(ctx._?.['VALID-AT']) ?? '9999-11-11T22:22:22.000Z',
-          )
-        },
-      })
-      // Run BEGIN
+      await this.set(new SessionContext(ctx))
       await this.send('BEGIN')
     } catch (e) {
-      this._release(dbc)
+      this.release()
       throw e
     }
     return this
   }
 
+  /**
+   * Commits a transaction and releases the connection to the pool.
+   */
   async commit() {
-    const dbc = this.dbc
-    if (!dbc) return
+    if (!this.dbc) return
     await this.send('COMMIT')
-    this._release(dbc) // only release on successful commit as otherwise released on rollback
+    this.release() // only release on successful commit as otherwise released on rollback
   }
 
+  /**
+   * Rolls back a transaction and releases the connection to the pool.
+   */
   async rollback() {
-    const dbc = this.dbc
-    if (!dbc) return
-    try {
-      await this.send('ROLLBACK')
-    } finally {
-      this._release(dbc)
-    }
+    if (!this.dbc) return
+    else
```

Comment:
no need for the `else`

Reply:
Yes, I sometimes do this for linguistic reasons and better readability (for  ;me)
We can remove it if it has the opposite effect on you, of course